plugins/shell_run/shell_run.7
.TH YED-SHELL_RUN 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
shell_run \- Run a shell command in the background and collect its output in the special buffer *shell-output.
.SH CONFIGURATION
None
.SH COMMANDS
.SS shell-run <command>
Run <command> in the background and focus *shell-output.
.SH BUFFERS
*shell-output
.SH NOTES
.P
Interactive commands should not be run with this command because the command will not receive any input.
Additionally, terminal escape codes in the command's output will not be interpreted as such.
.SH VERSION
0.0.1
.SH KEYWORDS
shell, run, command, bash, program, background


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


plugins/shell_run.c
#include <yed/plugin.h>

static yed_event_handler  pump_handler;
static int                cmd_is_running;
static yed_nb_subproc_t   nb_subproc;
static char              *cmd_string;

static yed_buffer * get_or_make_buffer(void);

static void shell_run_unload(yed_plugin *self);
static void shell_run_pump_handler(yed_event *event);

static void shell_run(int n_args, char **args);
static void shell_run_silent(int n_args, char **args);
static void shell_view_output(int n_args, char **args);

int yed_plugin_boot(yed_plugin *self) {
    YED_PLUG_VERSION_CHECK();

    get_or_make_buffer();

    yed_plugin_set_command(self, "shell-run",         shell_run);
    yed_plugin_set_command(self, "shell-run-silent",  shell_run_silent);
    yed_plugin_set_command(self, "shell-view-output", shell_view_output);

    pump_handler.kind = EVENT_PRE_PUMP;
    pump_handler.fn   = shell_run_pump_handler;

    yed_plugin_add_event_handler(self, pump_handler);

    yed_plugin_set_unload_fn(self, shell_run_unload);

    return 0;
}

static yed_buffer * get_or_make_buffer(void) {
    yed_buffer *buff;

    buff = yed_get_buffer("*shell-output");
    if (buff == NULL) {
        buff = yed_create_buffer("*shell-output");
        buff->flags |= BUFF_SPECIAL | BUFF_RD_ONLY;
    }

    return buff;
}

static void shell_run_unload(yed_plugin *self) {
    yed_buffer *buff;

    buff = yed_get_buffer("*shell-output");
    if (buff != NULL) {
        yed_free_buffer(buff);
    }

    if (cmd_string != NULL) {
        free(cmd_string);
        cmd_string = NULL;
    }
}

static void shell_run_update(void) {
    yed_frame **fit;
    int         last_row;

LOG_CMD_ENTER("shell-run");

    get_or_make_buffer()->flags &= ~BUFF_RD_ONLY;
    cmd_is_running = yed_read_subproc_into_buffer_nb(&nb_subproc);
    get_or_make_buffer()->flags |= BUFF_RD_ONLY;

    if (!cmd_is_running) {
        if (nb_subproc.err) {
            yed_cerr("something went wrong -- errno = %d\n", nb_subproc.err);
        } else {
            if (nb_subproc.exit_status != 0) {
                yed_cerr("'%s' failed with error code %d", cmd_string, nb_subproc.exit_status);
            }
        }
        free(cmd_string);
        cmd_string = NULL;
    }

    array_traverse(ys->frames, fit) {
        if (*fit == ys->active_frame) { continue; }
        if ((*fit)->buffer == get_or_make_buffer()) {
            last_row = yed_buff_n_lines((*fit)->buffer);
            yed_set_cursor_far_within_frame(*fit, last_row, 1);
        }
    }

LOG_EXIT();
}

static void shell_run_pump_handler(yed_event *event) {
    if (cmd_is_running) {
        shell_run_update();
    }
}

static void do_shell_run(int n_args, char **args) {
    array_t     string_build;
    const char *lazy_space;
    int         i;
    char       *full_cmd;

    string_build = array_make(char);

    lazy_space = "";
    for (i = 0; i < n_args; i += 1) {
        array_push_n(string_build, (char*)lazy_space, strlen(lazy_space));
        array_push_n(string_build, args[i], strlen(args[i]));
        lazy_space = " ";
    }

    array_zero_term(string_build);

    full_cmd = malloc(array_len(string_build) + 16);
    snprintf(full_cmd, array_len(string_build) + 16,
             "(%s) 2>&1", (char*)array_data(string_build));


    get_or_make_buffer()->flags &= ~BUFF_RD_ONLY;

    if (yed_start_read_subproc_into_buffer_nb(full_cmd, get_or_make_buffer(), &nb_subproc)) {
        get_or_make_buffer()->flags |= BUFF_RD_ONLY;

        yed_cerr("there was an error when calling yed_start_read_subproc_into_buffer_nb()");
        free(full_cmd);
        array_free(string_build);
        return;
    }

    get_or_make_buffer()->flags |= BUFF_RD_ONLY;

    free(full_cmd);

    cmd_string      = array_data(string_build);
    cmd_is_running  = 1;

    /*
    ** Wait 15 milliseconds so that the subprocess has time to start and
    ** do _something_.
    ** This makes it so that we can start to produce output before the next
    ** pump (which could be ~300 ms if there aren't any keypresses).
    */
    usleep(15000);
    shell_run_update();
}

static void shell_run(int n_args, char **args) {
    if (cmd_is_running) {
        yed_cerr("a command is already running!");
        return;
    }

    if (n_args == 0) {
        yed_cerr("expected at least 1 argument, but got %d", n_args);
        return;
    }

    do_shell_run(n_args, args);
    YEXE("special-buffer-prepare-focus", "*shell-output");
    yed_set_cursor_far_within_frame(ys->active_frame, 1, 1);
    YEXE("buffer", "*shell-output");
}

static void shell_run_silent(int n_args, char **args) {
    if (cmd_is_running) {
        yed_cerr("a command is already running!");
        return;
    }

    if (n_args == 0) {
        yed_cerr("expected at least 1 argument, but got %d", n_args);
        return;
    }

    do_shell_run(n_args, args);
}

static void shell_view_output(int n_args, char **args) {
    if (n_args != 0) {
        yed_cerr("expected 0 arguments, but got %d", n_args);
        return;
    }

    YEXE("special-buffer-prepare-focus", "*shell-output");
    yed_set_cursor_far_within_frame(ys->active_frame, 1, 1);
    YEXE("buffer", "*shell-output");
}
